include/ginkgo/core/config/data.hpp

```diff
@@ -0,0 +1,242 @@
+ */
+class data {
+    template <typename T>
+    friend T get(const data& d);
+
+    template <typename T>
+    friend bool holds_alternative(const data& d);
+
+public:
+    /**
+     * Default Constructor
+     */
+    data() : tag_(tag_type::empty_t) {}
+
+    /**
+     * Constructor for bool
+     */
+    data(bool bb) : tag_(tag_type::bool_t) { u_.bool_ = bb; }
+
+    /**
+     * Constructor for integer
+     */
+    data(long long int ii) : tag_(tag_type::int_t) { u_.int_ = ii; }
+
+    /**
+     * Constructor for integer with all integer type except for unsigned long
+     * long int
+     */
+    template <typename T,
+              typename = typename std::enable_if<
+                  std::is_integral<T>::value &&
+                  !std::is_same<T, unsigned long long int>::value>::type>
+    data(T ii) : data(static_cast<long long int>(ii))
+    {}
+
+    /**
+     * Constructor for string
+     */
+    data(const std::string& str) : tag_(tag_type::str_t) { str_ = str; };
+
+    /**
+     * Constructor for char (otherwise, it will use bool)
+     */
+    data(const char* s) : data(std::string(s)) {}
+
+    /**
+     * Constructor for double
+     */
+    data(double dd) : tag_(tag_type::double_t) { u_.double_ = dd; };
+
+    /**
+     * Constructor for float
+     */
+    data(float dd) : data(static_cast<double>(dd)) {}
+
+protected:
+    enum tag_type { str_t, int_t, double_t, bool_t, empty_t };
+
+    template <tag_type>
+    struct tag {
+        using type = void;
+    };
+
+    /**
+     * Get the data with Type T
+     */
+    template <typename T>
+    inline T get() const
+    {
+        throw std::runtime_error("Not Supported");
+    }
+
+    template <tag_type T>
+    inline typename tag<T>::type get() const;
+
+    inline tag_type get_tag() const { return tag_; }
+
+private:
+    tag_type tag_;
+    std::string str_;
+    union {
+        long long int int_;
+        double double_;
+        bool bool_;
+    } u_;
+};
+
+template <>
+inline long long int data::get<long long int>() const
+{
+    assert(tag_ == data::tag_type::int_t);
+    return u_.int_;
+}
+
+template <>
+inline std::string data::get<std::string>() const
+{
+    assert(tag_ == data::tag_type::str_t);
+    return str_;
+}
+
+template <>
+inline double data::get<double>() const
+{
+    assert(tag_ == data::tag_type::double_t);
+    return u_.double_;
+}
+
+template <>
+inline bool data::get<bool>() const
+{
+    assert(tag_ == data::tag_type::bool_t);
+    return u_.bool_;
+}
+
+
+template <>
+struct data::tag<data::tag_type::bool_t> {
+    using type = bool;
+};
+
+template <>
+struct data::tag<data::tag_type::int_t> {
+    using type = long long int;
+};
+
+template <>
+struct data::tag<data::tag_type::double_t> {
+    using type = double;
+};
+
+template <>
+struct data::tag<data::tag_type::str_t> {
+    using type = std::string;
+};
+
+
+template <typename T>
+inline T get(const data& d)
+{
+    static_assert(std::is_same<T, std::string>::value ||
```

Comment:
it is to provide the same rule as std::variant.
I did the get with different type in the config helper.

Reply:
As I mentioned before, I don't think we should aim for `std::variant` compatability here. This type should not be exposed to the user, so we can easily replace it later on.

Reply:
An alternative would be having `get_string`, `get_int`, `get_double` functions next to `get_array`, that would also get rid of the need to use templates. I don't think we need these in a generic context anyways, a solver would always query specific properties by type and name.

Reply:
I think that is a good option. But I wouldn't use `int` or `double` in the name, since at least `int` is just wrong. So perhaps we could have `get_integer` and `get_real` instead. Then it would be a bit clearer that additional `static_cast` are necessary.

Reply:
For me, the template already did the same thing. You get what you require if it is legal. Query specific properties without the template may make the following pr a bit messy. I can try these first to see what changes necessary in the following pr.